rts/Lua/LuaMatrix.hpp

```diff
@@ -0,0 +1,197 @@
+
+	bool FeatureMatrix(const unsigned int featureID, const sol::optional<bool> mult, sol::this_state L) { return ObjectMatImpl<CFeature>(featureID, mult.value_or(objectMultDefault), L); }
+	bool FeaturePieceMatrix(const unsigned int featureID, const unsigned int pieceNum, const sol::optional<bool> mult, sol::this_state L) { return ObjectPieceMatImpl<CFeature>(featureID, pieceNum, mult.value_or(objectMultDefault), L); }
+
+	bool ProjectileMatrix(const unsigned int projID, const sol::optional<bool> mult, sol::this_state L) { return ObjectMatImpl<CFeature>(projID, mult.value_or(objectMultDefault), L); }
+	bool ProjectilePieceMatrix(const unsigned int projID, const unsigned int pieceNum, const sol::optional<bool> mult, sol::this_state L) { return ObjectPieceMatImpl<CFeature>(projID, pieceNum, mult.value_or(objectMultDefault), L); } //does this exist at all?
+
+	void ScreenViewMatrix(const sol::optional<bool> mult);
+	void ScreenProjMatrix(const sol::optional<bool> mult);
+
+	void SunViewMatrix(const sol::optional<bool> mult);
+	void SunProjMatrix(const sol::optional<bool> mult);
+
+	void Ortho(const float left, const float right, const float bottom, const float top, const float near, const float far, const sol::optional<bool> mult);
+	void Frustum(const float left, const float right, const float bottom, const float top, const float near, const float far, const sol::optional<bool> mult);
+	void Billboard(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+
+	tuple16f GetAsScalar()
+	{
+		return std::make_tuple(
+			mat[0], mat[1], mat[2], mat[3],
+			mat[4], mat[5], mat[6], mat[7],
+			mat[8], mat[9], mat[10], mat[11],
+			mat[12], mat[13], mat[14], mat[15]
+		);
+	}
+
+	sol::as_table_t<std::vector<float>> GetAsTable()
+	{
+		return sol::as_table(std::vector<float> {
+			mat[0], mat[1], mat[2], mat[3],
+			mat[4], mat[5], mat[6], mat[7],
+			mat[8], mat[9], mat[10], mat[11],
+			mat[12], mat[13], mat[14], mat[15]
+		});
+	}
+
+public:
+	const CMatrix44f& GetMatRef() const { return  mat; }
+	const CMatrix44f* GetMatPtr() const { return &mat; }
+public:
+	sol::as_table_t<std::vector<float>> operator* (const sol::table& tbl) const {
+		float4 f4 { 0.0f, 0.0f, 0.0f, 1.0f };
+		for (auto i = 1; i <= 4; ++i) {
+			const sol::optional<float> maybeValue = tbl[i];
+			if (!maybeValue)
+				continue;
+
+			f4[i - 1] = maybeValue.value();
+		}
+		f4 = mat * f4;
+		return sol::as_table( std::vector<float> { f4.x, f4.y, f4.z, f4.w } );
+	};
+	LuaMatrixImpl operator* (const LuaMatrixImpl& lmi) const { return LuaMatrixImpl(mat * lmi.GetMatRef()); };
+
+	LuaMatrixImpl operator+ (const LuaMatrixImpl& lmi) const { return LuaMatrixImpl(mat + lmi.GetMatRef()); };
+private:
+	template<typename TFunc>
+	inline void AssignOrMultMatImpl(const sol::optional<bool> mult, const bool multDefault, const TFunc& tfunc);
```

Comment:
it makes little sense to declare these functions here seeing as they can't be invoked (templates need to be fully defined to be invoked). Most of these templates can be free functions, either static or in a namespace